sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc

```diff
@@ -303,6 +315,281 @@ int main(int argc, char *argv[]) {
 }
 ----
 
+=== Math Functions
+
+The following functions are only available when `T` is `bfloat16` or
+`sycl::marray<bfloat16, {N}>`, where `{N}` means any positive value of
+`size_t` type.
+
+==== isnan
+
+```c++
+namespace sycl::ext::oneapi {
+
+bool isnan(bfloat16 x);
+
+template <size_t N>
+sycl::marray<bool, N> isnan(sycl::marray<bfloat16, N> x);
+} // namespace sycl::ext::oneapi
```

Comment:
All makes sense to me. For the vec question. I don't think it was excluded. I didn't add it initially when creating the first iteration of the bfloat16 math function extension because we did not have a reason to add vec when we had marray in the cuda backend, but I imagined that at some point other backends would motivate the addition of vec math functions.
Even if they are only add them purely to have math functions for scalar, vec, and marray types that are consistent with the rest of the types, I see that this could make sense.
I can make such a change to the spec, and add a generic implementation for bfloat16 math functions for vec type in this PR?
Or perhaps I could delay this till later if you think that is better?
I suppose features can still be added to supported extensions so long as they are new additions rather than changes to existing functions/types etc?

Reply:
After reflection, I think it would make sense to wait until the revision 8 is published, and then use the new math function definition patterns for the bfloat16 math functions too.

Since bfloat16 math functions are new, would it be better to just support them using the new math function definition patterns (forms)? As far as I understand it the move to the new form doesn't actually change how a user would define the function in practice, just that a mixture of overloading and templating is used instead of the previous situation that just used templates?

If the above is the case it would seem overly complex to introduce multiple macros for bfloat16 math functions, especially if they form part of the extension for the bfloat16 class.

Reply:
> After reflection, I think it would make sense to wait until the revision 8 is published, and then use the new math function definition patterns for the bfloat16 math functions too.

I expect revision 8 to be published soon.  Yes, I think it would be good for this extension specification to use the same pattern and specification style that revision 8 has.

> Since bfloat16 math functions are new, would it be better to just support them using the new math function definition patterns (forms)? As far as I understand it the move to the new form doesn't actually change how a user would define the function in practice, just that the new forms use a mixture of overloading and templating instead of the previous situation that just used templates?

Most application code won't be impacted by the move from old to new form, but there are some edge cases that will be visible to applications.  I think I may have been premature about the macro.  Our plan is to have a macro in the headers, which selects between the old and new forms of the builtin functions, but I think it is not implemented yet.

I agree that in theory the support for bfloat16 could be added with *only* the new form (not selected by the macro) because we have no legacy code to support with the old form.  I was just thinking that it might be hard to implement this way.  I assume the current implementation of the math functions has a set of constraints for the `NonScalar` template parameter.  Therefore, I was thinking that you would simply broaden these constraints to allow `NonScalar` to be `marray<bfloat16>`.  Once we add the macro, wouldn't this mean that the `bfloat16` support was also affected by the macro?

Reply:
I can only answer this question:

> Is the long term plan that bfloat16 type and associated functions be a core part of sycl, or available as a header only library?

Yes, I think we should consider eventually adding bfloat16 as a core part of SYCL.  We don't want it to be a header-only library because it needs some underlying h/w support for efficiency.

We need someone to be the "owner" of the bfloat16 feature, who can decide what functions make sense to add to this extension.